pkg/asset/quota/openstack/openstack.go

```diff
@@ -11,44 +14,90 @@ import (
 // These constraints can be used to check if there is enough quota for creating a cluster
 // for the install config.
 func Constraints(ci *validation.CloudInfo, controlPlanes []machineapi.Machine, computes []machineapi.MachineSet) []quota.Constraint {
-	ctrplConfigs := make([]*openstackprovider.OpenstackProviderSpec, len(controlPlanes))
-	for i, m := range controlPlanes {
-		ctrplConfigs[i] = m.Spec.ProviderSpec.Value.Object.(*openstackprovider.OpenstackProviderSpec)
+	var constraints []quota.Constraint
+
+	for i := 0; i < len(controlPlanes); i++ {
+		constraints = append(constraints, machineConstraints(ci, &controlPlanes[i])...)
+	}
+	constraints = append(constraints, instanceConstraint(int64(len(controlPlanes))))
+
+	for i := 0; i < len(computes); i++ {
+		constraints = append(constraints, machineSetConstraints(ci, &computes[i])...)
+	}
+	constraints = append(constraints, instanceConstraint(int64(len(computes))))
+
+	return aggregate(constraints)
+}
+
+func getOpenstackProviderSpec(spec *machineapi.ProviderSpec) *openstackprovider.OpenstackProviderSpec {
+	// An invalid ProviderSpec would presumably be a programming error here.
+	if spec.Value == nil {
+		logrus.Warnf("Empty ProviderSpec")
+		return nil
 	}
-	computeReplicas := make([]int64, len(computes))
-	computeConfigs := make([]*openstackprovider.OpenstackProviderSpec, len(computes))
-	for i, m := range computes {
-		computeReplicas[i] = int64(*m.Spec.Replicas)
-		computeConfigs[i] = m.Spec.Template.Spec.ProviderSpec.Value.Object.(*openstackprovider.OpenstackProviderSpec)
+
+	osps, ok := spec.Value.Object.(*openstackprovider.OpenstackProviderSpec)
+	if !ok {
+		logrus.Warnf("ProviderSpec of unexpected type %T", spec.Value.Object)
+		return nil
 	}
 
-	var ret []quota.Constraint
-	ret = controlPlane(ci, ctrplConfigs)
-	ret = append(ret, compute(ci, computeReplicas, computeConfigs)...)
-	return aggregate(ret)
+	return osps
 }
 
-func controlPlane(ci *validation.CloudInfo, machines []*openstackprovider.OpenstackProviderSpec) []quota.Constraint {
-	var ret []quota.Constraint
-	for _, m := range machines {
-		flavorInfo := ci.Flavors[m.Flavor]
-		ret = append(ret, machineFlavorCoresToQuota(flavorInfo, ci), machineFlavorRAMToQuota(flavorInfo, ci))
+func getFlavor(ci *validation.CloudInfo, flavorName string) *flavors.Flavor {
+	if flavorInfo, ok := ci.Flavors[flavorName]; ok {
+		return flavorInfo.Flavor
 	}
-	ret = append(ret, instanceConstraint(int64(len(machines))))
-	return ret
+	return nil
```

Comment:
Are you trying to fix Go? Please don't :laughing:

Reply:
This is subtle, but on closer inspection I remember why I did this: flavorInfo.Flavor is also a pointer so therefore may *also* be nil. i.e. checking `ok` here is insufficient to ensure you don't have a nil pointer.

Honestly I think this is an unnecessary landmine in this datastructure, so I'm going to try to change that in another small patch to see what you think.